File: packages/python/src/agent_venv/registry.py
"""Persistent environment registry.

See spec/registry.md.
"""

from __future__ import annotations

import dataclasses
import datetime as _dt
import hashlib
import json
import os
import shutil
import time
from pathlib import Path
from typing import Any

from . import errors as _errors


REGISTRY_SCHEMA_VERSION = 1


def default_registry_root() -> Path:
    """Resolve the default registry path per the spec."""

    env_override = os.environ.get("AGENT_VENV_REGISTRY_ROOT")
    if env_override:
        return Path(env_override)
    xdg = os.environ.get("XDG_DATA_HOME")
    if xdg:
        return Path(xdg) / "agent-venv" / "envs"
    return Path.home() / ".local" / "share" / "agent-venv" / "envs"


def slug_for(name: str) -> str:
    return hashlib.sha256(name.encode("utf-8")).hexdigest()[:16]


@dataclasses.dataclass
class IndexEntry:
    name: str
    relative_dir: str  # relative to registry_root


@dataclasses.dataclass
class Metadata:
    schema_version: int
    name: str
    adapter_id: str
    created_at: str
    env_overrides: dict[str, str]
    credentials_loaded: bool
    credentials_loaded_at: str | None

    def to_dict(self) -> dict[str, Any]:
        return dataclasses.asdict(self)

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> Metadata:
        return cls(
            schema_version=int(data.get("schema_version", 1)),
            name=str(data["name"]),
            adapter_id=str(data["adapter_id"]),
            created_at=str(data.get("created_at", "")),
            env_overrides=dict(data.get("env_overrides") or {}),
            credentials_loaded=bool(data.get("credentials_loaded", False)),
            credentials_loaded_at=data.get("credentials_loaded_at"),
        )


class Registry:
    """File-system-backed registry of persistent environments.

    Locking is advisory via a simple lock file + ``fcntl.flock`` on Unix.
    """

    def __init__(self, root: Path):
        self.root = root.resolve() if root.exists() else root.absolute()
        self._index_path = self.root / "index.json"
        self._envs_dir = self.root / "envs"
        self._lock_path = self.root / ".lock"

    # ------------------------------------------------------------------
    # Locking
    # ------------------------------------------------------------------

    def _acquire_lock(self) -> Any:
        self.root.mkdir(parents=True, exist_ok=True)
        # Lazy import: only Unix has fcntl. On Windows we don't lock in v0.
        try:
            import fcntl

            f = open(self._lock_path, "w")
            for _ in range(50):
                try:
                    fcntl.flock(f.fileno(), fcntl.LOCK_EX | fcntl.LOCK_NB)
                    return f
                except OSError:
                    time.sleep(0.1)
            raise _errors.RegistryUnavailableError(
                "could not acquire registry lock", path=str(self._lock_path)
            )
        except ImportError:
            return open(self._lock_path, "w")

    def _release_lock(self, fp: Any) -> None:
        try:
            import fcntl

            fcntl.flock(fp.fileno(), fcntl.LOCK_UN)
        except (ImportError, OSError):
            pass
        try:
            fp.close()
        except OSError:
            pass

    # ------------------------------------------------------------------
    # Index
    # ------------------------------------------------------------------

    def _read_index(self) -> dict[str, str]:
        if not self._index_path.exists():
            return {}
        try:
            data = json.loads(self._index_path.read_text(encoding="utf-8"))
        except (OSError, json.JSONDecodeError) as exc:
            raise _errors.RegistryUnavailableError(
                f"reading index: {exc}", path=str(self._index_path)
            ) from exc
        entries = data.get("entries") or {}
        return {str(k): str(v) for k, v in entries.items()}

    def _write_index(self, entries: dict[str, str]) -> None:
        payload = {
            "schema_version": REGISTRY_SCHEMA_VERSION,
            "entries": dict(entries),
        }
        self.root.mkdir(parents=True, exist_ok=True)
        tmp = self._index_path.with_suffix(".json.tmp")
        tmp.write_text(json.dumps(payload, indent=2), encoding="utf-8")
        os.replace(tmp, self._index_path)

    # ------------------------------------------------------------------
    # Public ops
    # ------------------------------------------------------------------

    def list_names(self) -> list[str]:
        return sorted(self._read_index().keys())

    def lookup(self, name: str) -> tuple[Path, Metadata] | None:
        entries = self._read_index()
        rel = entries.get(name)
        if rel is None:
            return None
        env_dir = (self.root / rel).resolve() if (self.root / rel).exists() else self.root / rel
        meta_path = env_dir / "metadata.json"
        if not meta_path.exists():
            return None
        try:
            meta = Metadata.from_dict(json.loads(meta_path.read_text(encoding="utf-8")))
        except (OSError, json.JSONDecodeError, KeyError) as exc:
            raise _errors.RegistryUnavailableError(
                f"reading metadata for {name!r}: {exc}", path=str(meta_path)
            ) from exc
        return env_dir, meta

    def reserve_or_get(
        self, name: str, adapter_id: str
    ) -> tuple[Path, Metadata, bool]:
        """Returns ``(env_dir, metadata, created)``.

        If the name is new, allocates a fresh env_dir, writes a stub metadata,
        and registers it in the index. If the name already exists, validates
        adapter_id against the recorded one (raising AdapterMismatchError on
        disagreement) and returns the existing entry without modification.
        """

        lock = self._acquire_lock()
        try:
            entries = self._read_index()
            if name in entries:
                rel = entries[name]
                env_dir = (self.root / rel).resolve() if (self.root / rel).exists() else self.root / rel
                meta_path = env_dir / "metadata.json"
                meta = Metadata.from_dict(json.loads(meta_path.read_text(encoding="utf-8")))
                if meta.adapter_id != adapter_id:
                    raise _errors.AdapterMismatchError(
                        f"env {name!r} was created with adapter {meta.adapter_id!r} "
                        f"but {adapter_id!r} was requested",
                        name=name,
                        expected_adapter_id=meta.adapter_id,
                        actual_adapter_id=adapter_id,
                    )
                return env_dir, meta, False
            slug = slug_for(name)
            # Avoid slug collision (vanishingly unlikely with sha256, but be defensive).
            existing_slugs = {Path(p).name for p in entries.values()}
            attempt = slug
            i = 0
            while attempt in existing_slugs:
                i += 1
                attempt = f"{slug}-{i}"
            slug = attempt
            rel_dir = f"envs/{slug}"
            env_dir = (self.root / rel_dir).absolute()
            (env_dir / "profile").mkdir(parents=True, exist_ok=True)
            now = _dt.datetime.now(_dt.timezone.utc).isoformat()
            meta = Metadata(
                schema_version=REGISTRY_SCHEMA_VERSION,
                name=name,
                adapter_id=adapter_id,
                created_at=now,
                env_overrides={},
                credentials_loaded=False,
                credentials_loaded_at=None,
            )
            (env_dir / "metadata.json").write_text(
                json.dumps(meta.to_dict(), indent=2), encoding="utf-8"
            )
            entries[name] = rel_dir
            self._write_index(entries)
            return env_dir.resolve(), meta, True
        finally:
            self._release_lock(lock)

    def update_metadata(self, env_dir: Path, meta: Metadata) -> None:
        path = env_dir / "metadata.json"
        try:
            path.write_text(json.dumps(meta.to_dict(), indent=2), encoding="utf-8")
        except OSError as exc:
            raise _errors.RegistryUnavailableError(
                f"writing metadata: {exc}", path=str(path)
            ) from exc

    def remove(self, name: str) -> tuple[bool, Path | None, str | None]:
        """Returns ``(ok, env_dir, error_message)``.

        Removes the on-disk env dir and the index entry. If the on-disk
        removal fails, the index entry is still removed and ok=False with a
        message.
        """

        lock = self._acquire_lock()
        try:
            entries = self._read_index()
            if name not in entries:
                raise _errors.EnvironmentNotFoundError(
                    f"no environment named {name!r}", name=name, registry_root=str(self.root)
                )
            rel = entries.pop(name)
            env_dir = (self.root / rel).absolute()
            cleanup_err: str | None = None
            try:
                if env_dir.exists():
                    shutil.rmtree(env_dir)
            except OSError as exc:
                cleanup_err = str(exc)
            self._write_index(entries)
            return cleanup_err is None, env_dir, cleanup_err
        finally:
            self._release_lock(lock)

File: packages/rust/src/registry.rs
use crate::error::{Error, Result};
use chrono::Utc;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use sha2::{Digest, Sha256};
use std::collections::HashMap;
use std::fs::{File, OpenOptions};
use std::io::ErrorKind as IoErrorKind;
use std::path::{Path, PathBuf};

pub const REGISTRY_SCHEMA_VERSION: u32 = 1;

pub fn default_registry_root() -> PathBuf {
    if let Ok(s) = std::env::var("AGENT_VENV_REGISTRY_ROOT") {
        return PathBuf::from(s);
    }
    if let Ok(s) = std::env::var("XDG_DATA_HOME") {
        return PathBuf::from(s).join("agent-venv").join("envs");
    }
    let home = std::env::var_os("HOME")
        .map(PathBuf::from)
        .unwrap_or_else(|| PathBuf::from("/"));
    home.join(".local").join("share").join("agent-venv").join("envs")
}

pub fn slug_for(name: &str) -> String {
    let mut h = Sha256::new();
    h.update(name.as_bytes());
    let digest = h.finalize();
    hex::encode(&digest[..8])
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Metadata {
    pub schema_version: u32,
    pub name: String,
    pub adapter_id: String,
    pub created_at: String,
    #[serde(default)]
    pub env_overrides: HashMap<String, String>,
    #[serde(default)]
    pub credentials_loaded: bool,
    #[serde(default)]
    pub credentials_loaded_at: Option<String>,
}

pub struct Registry {
    pub root: PathBuf,
    pub index_path: PathBuf,
    pub envs_dir: PathBuf,
    pub lock_path: PathBuf,
}

impl Registry {
    pub fn new(root: impl Into<PathBuf>) -> Self {
        let root = root.into();
        let index_path = root.join("index.json");
        let envs_dir = root.join("envs");
        let lock_path = root.join(".lock");
        Self {
            root,
            index_path,
            envs_dir,
            lock_path,
        }
    }

    fn acquire_lock(&self) -> Result<File> {
        std::fs::create_dir_all(&self.root).map_err(|e| Error::RegistryUnavailable {
            reason: e.to_string(),
            path: self.root.display().to_string(),
        })?;
        for _ in 0..200 {
            match OpenOptions::new()
                .write(true)
                .create_new(true)
                .open(&self.lock_path)
            {
                Ok(f) => return Ok(f),
                Err(e) if e.kind() == IoErrorKind::AlreadyExists => {
                    std::thread::sleep(std::time::Duration::from_millis(25));
                }
                Err(e) => {
                    return Err(Error::RegistryUnavailable {
                        reason: e.to_string(),
                        path: self.lock_path.display().to_string(),
                    });
                }
            }
        }
        Err(Error::RegistryUnavailable {
            reason: "could not acquire lock".into(),
            path: self.lock_path.display().to_string(),
        })
    }

    fn release_lock(&self, _f: File) {
        let _ = std::fs::remove_file(&self.lock_path);
    }

    fn read_index(&self) -> Result<HashMap<String, String>> {
        match std::fs::read_to_string(&self.index_path) {
            Ok(s) => {
                let v: Value =
                    serde_json::from_str(&s).map_err(|e| Error::RegistryUnavailable {
                        reason: format!("parse index: {e}"),
                        path: self.index_path.display().to_string(),
                    })?;
                let entries = v.get("entries").and_then(Value::as_object);
                let mut out = HashMap::new();
                if let Some(o) = entries {
                    for (k, v) in o {
                        if let Some(s) = v.as_str() {
                            out.insert(k.clone(), s.to_string());
                        }
                    }
                }
                Ok(out)
            }
            Err(e) if e.kind() == IoErrorKind::NotFound => Ok(HashMap::new()),
            Err(e) => Err(Error::RegistryUnavailable {
                reason: e.to_string(),
                path: self.index_path.display().to_string(),
            }),
        }
    }

    fn write_index(&self, entries: &HashMap<String, String>) -> Result<()> {
        std::fs::create_dir_all(&self.root).map_err(|e| Error::RegistryUnavailable {
            reason: e.to_string(),
            path: self.root.display().to_string(),
        })?;
        let mut payload = serde_json::Map::new();
        payload.insert(
            "schema_version".into(),
            Value::from(REGISTRY_SCHEMA_VERSION),
        );
        let mut entries_obj = serde_json::Map::new();
        for (k, v) in entries {
            entries_obj.insert(k.clone(), Value::from(v.clone()));
        }
        payload.insert("entries".into(), Value::Object(entries_obj));
        let serialized = serde_json::to_string_pretty(&Value::Object(payload)).unwrap();
        let tmp = self.index_path.with_extension("json.tmp");
        std::fs::write(&tmp, serialized).map_err(|e| Error::RegistryUnavailable {
            reason: e.to_string(),
            path: tmp.display().to_string(),
        })?;
        std::fs::rename(&tmp, &self.index_path).map_err(|e| Error::RegistryUnavailable {
            reason: e.to_string(),
            path: self.index_path.display().to_string(),
        })?;
        Ok(())
    }

    pub fn list_names(&self) -> Result<Vec<String>> {
        let mut names: Vec<String> = self.read_index()?.into_keys().collect();
        names.sort();
        Ok(names)
    }

    pub fn lookup(&self, name: &str) -> Result<Option<(PathBuf, Metadata)>> {
        let entries = self.read_index()?;
        let rel = match entries.get(name) {
            Some(r) => r.clone(),
            None => return Ok(None),
        };
        let env_dir = std::fs::canonicalize(self.root.join(&rel))
            .unwrap_or_else(|_| self.root.join(&rel));
        let meta_path = env_dir.join("metadata.json");
        match std::fs::read_to_string(&meta_path) {
            Ok(s) => {
                let meta: Metadata =
                    serde_json::from_str(&s).map_err(|e| Error::RegistryUnavailable {
                        reason: format!("parse metadata: {e}"),
                        path: meta_path.display().to_string(),
                    })?;
                Ok(Some((env_dir, meta)))
            }
            Err(_) => Ok(None),
        }
    }

    pub fn reserve_or_get(
        &self,
        name: &str,
        adapter_id: &str,
    ) -> Result<(PathBuf, Metadata, bool)> {
        let lock = self.acquire_lock()?;
        let result = (|| -> Result<(PathBuf, Metadata, bool)> {
            let mut entries = self.read_index()?;
            if let Some(rel) = entries.get(name) {
                let env_dir = std::fs::canonicalize(self.root.join(rel))
                    .unwrap_or_else(|_| self.root.join(rel));
                let meta_path = env_dir.join("metadata.json");
                let meta: Metadata = serde_json::from_str(
                    &std::fs::read_to_string(&meta_path).map_err(|e| {
                        Error::RegistryUnavailable {
                            reason: e.to_string(),
                            path: meta_path.display().to_string(),
                        }
                    })?,
                )
                .map_err(|e| Error::RegistryUnavailable {
                    reason: format!("parse metadata: {e}"),
                    path: meta_path.display().to_string(),
                })?;
                if meta.adapter_id != adapter_id {
                    return Err(Error::AdapterMismatch {
                        name: name.to_string(),
                        expected_adapter_id: meta.adapter_id.clone(),
                        actual_adapter_id: adapter_id.to_string(),
                    });
                }
                return Ok((env_dir, meta, false));
            }
            let mut slug = slug_for(name);
            let existing: std::collections::HashSet<String> = entries
                .values()
                .map(|p| Path::new(p).file_name().unwrap().to_string_lossy().to_string())
                .collect();
            let mut attempt = slug.clone();
            let mut i = 0;
            while existing.contains(&attempt) {
                i += 1;
                attempt = format!("{slug}-{i}");
            }
            slug = attempt;
            let rel_dir = format!("envs/{slug}");
            let env_dir_initial = self.root.join(&rel_dir);
            std::fs::create_dir_all(env_dir_initial.join("profile")).map_err(|e| {
                Error::RegistryUnavailable {
                    reason: e.to_string(),
                    path: env_dir_initial.display().to_string(),
                }
            })?;
            let env_dir = std::fs::canonicalize(&env_dir_initial)
                .unwrap_or_else(|_| env_dir_initial.clone());
            let meta = Metadata {
                schema_version: REGISTRY_SCHEMA_VERSION,
                name: name.to_string(),
                adapter_id: adapter_id.to_string(),
                created_at: Utc::now().to_rfc3339(),
                env_overrides: HashMap::new(),
                credentials_loaded: false,
                credentials_loaded_at: None,
            };
            std::fs::write(
                env_dir.join("metadata.json"),
                serde_json::to_string_pretty(&meta).unwrap(),
            )
            .map_err(|e| Error::RegistryUnavailable {
                reason: e.to_string(),
                path: env_dir.join("metadata.json").display().to_string(),
            })?;
            entries.insert(name.to_string(), rel_dir);
            self.write_index(&entries)?;
            Ok((env_dir, meta, true))
        })();
        self.release_lock(lock);
        result
    }

    pub fn update_metadata(&self, env_dir: &Path, meta: &Metadata) -> Result<()> {
        let path = env_dir.join("metadata.json");
        std::fs::write(&path, serde_json::to_string_pretty(meta).unwrap()).map_err(|e| {
            Error::RegistryUnavailable {
                reason: e.to_string(),
                path: path.display().to_string(),
            }
        })
    }

    pub fn remove(&self, name: &str) -> Result<(bool, PathBuf, Option<String>)> {
        let lock = self.acquire_lock()?;
        let result = (|| -> Result<(bool, PathBuf, Option<String>)> {
            let mut entries = self.read_index()?;
            let rel = match entries.remove(name) {
                Some(r) => r,
                None => {
                    return Err(Error::EnvironmentNotFound {
                        name: name.to_string(),
                        registry_root: self.root.display().to_string(),
                    });
                }
            };
            let env_dir = self.root.join(&rel);
            let mut cleanup_err: Option<String> = None;
            if env_dir.exists() {
                if let Err(e) = std::fs::remove_dir_all(&env_dir) {
                    cleanup_err = Some(e.to_string());
                }
            }
            self.write_index(&entries)?;
            Ok((cleanup_err.is_none(), env_dir, cleanup_err))
        })();
        self.release_lock(lock);
        result
    }
}
